### PR TITLE
Generic OAuth2 support - refs BT#16734

### PR DESCRIPTION
**Pull request by sebastiendu**

### CODE REVIEW

File: plugin/oauth2/install.php [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<?php
```
Comment (jmontoyaa) [inline plugin/oauth2/install.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

**Comment by jmontoyaa** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 9
           

Complexity increasing per file
==============================
- plugin/oauth2/src/OAuth2.php  11
- plugin/oauth2/config.dist.php  1
- main/inc/lib/online.inc.php  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956&bid=18593000)

File: plugin/oauth2/login.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
+<?php
```
Comment (jmontoyaa) [inline plugin/oauth2/login.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: main/auth/external_login/login.oauth2.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
+<?php
```
Comment (jmontoyaa) [inline main/auth/external_login/login.oauth2.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/index.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
+<?php
```
Comment (jmontoyaa) [inline plugin/oauth2/index.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/config.dist.php [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<?php
```
Comment (jmontoyaa) [inline plugin/oauth2/config.dist.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/src/OAuth2.php [hunk 6]
```diff
@@ -0,0 +1,383 @@
+<?php
```
Comment (jmontoyaa) [inline plugin/oauth2/src/OAuth2.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/lang/english.php [hunk 7]
```diff
@@ -0,0 +1,116 @@
+<?php
```
Comment (jmontoyaa) [inline plugin/oauth2/lang/english.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/src/callback.php [hunk 8]
```diff
@@ -0,0 +1,71 @@
+<?php
```
Comment (jmontoyaa) [inline plugin/oauth2/src/callback.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/plugin.php [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<?php
```
Comment (jmontoyaa) [inline plugin/oauth2/plugin.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

**Comment by jmontoyaa** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 9
           

Complexity increasing per file
==============================
- plugin/oauth2/src/OAuth2.php  11
- main/inc/lib/online.inc.php  3
- plugin/oauth2/config.dist.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956&bid=18593000)

File: plugin/oauth2/index.php (hunk 4 above)
Comment (jmontoyaa) [inline plugin/oauth2/index.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: main/auth/external_login/login.oauth2.php (hunk 3 above)
Comment (jmontoyaa) [inline main/auth/external_login/login.oauth2.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/config.dist.php (hunk 5 above)
Comment (jmontoyaa) [inline plugin/oauth2/config.dist.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/src/OAuth2.php [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
```
Comment (jmontoyaa) [inline plugin/oauth2/src/OAuth2.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/plugin.php (hunk 9 above)
Comment (jmontoyaa) [inline plugin/oauth2/plugin.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/install.php (hunk 1 above)
Comment (jmontoyaa) [inline plugin/oauth2/install.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/lang/english.php [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
```
Comment (jmontoyaa) [inline plugin/oauth2/lang/english.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/src/callback.php (hunk 8 above)
Comment (jmontoyaa) [inline plugin/oauth2/src/callback.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/login.php (hunk 2 above)
Comment (jmontoyaa) [inline plugin/oauth2/login.php:1]: Codacy Issue found: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

**Comment by NicoDucou** [comment]
Hi @jmontoyaa 
It seems all good to me. Can you review this PR to integrate it in 1.11.x ?

File: main/inc/lib/online.inc.php [hunk 12]
```diff
@@ -199,6 +199,19 @@ function online_logout($user_id = null, $logout_redirect = false)
         }
     }
 
+    if ('true' === api_get_plugin_setting('oauth2', 'enable')
+        and 'oauth2' === ChamiloSession::read('_user_auth_source')
+        and ChamiloSession::has('oauth2AccessToken')
+    ) {
```
Comment (jmontoyaa) [inline main/inc/lib/online.inc.php:205]: Use "&&" instead of "and"

File: plugin/oauth2/src/OAuth2.php [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+     *
+     * @param $userId integer
+     * @param $response array
+     */
+    private function updateUserUrls($userId, $response)
+    {
+        if (api_is_multiple_url_enabled()) {
+            $key = $this->get(self::SETTING_RESPONSE_RESOURCE_OWNER_URLS);
+            if (!empty($key)) {
+                require_once __DIR__.'/../../../main/inc/lib/urlmanager.lib.php';
```
Comment (jmontoyaa) [inline plugin/oauth2/src/OAuth2.php:351]: "urlmanager.lib.php" should be already loaded by composer, so no need to do this require_once

File: main/auth/external_login/login.oauth2.php [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+/* For licensing terms, see /license.txt */
+
+require_once __DIR__.'/functions.inc.php';
+
+/** @var array $uData */
+if ($uData['auth_source'] === 'oauth2') {
```
Comment (jmontoyaa) [inline main/auth/external_login/login.oauth2.php:7]: `if ($uData && 'oauth2' === $uData['auth_source']) {`

File: plugin/oauth2/config.dist.php [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+/**
+ * Adjust user account values.
+ *
+ * @param array                          $response the Resource Owner Details
+ * @param Chamilo\UserBundle\Entity\User $user     the user
+ *                                                 with firstname, lastname, status, username and email already updated
+ *                                                 but unsaved
+ */
+function oauth2_update_user_from_resource_owner_details(array $response, Chamilo\UserBundle\Entity\User $user)
```
Comment (jmontoyaa) [inline plugin/oauth2/config.dist.php:10]: use camelCase for the function name: 

`oauth2UpdateUserFromResourceOwnerDetails`

File: plugin/oauth2/src/OAuth2.php [hunk 16]
```diff
@@ -0,0 +1,382 @@
...
+        }
+        $extraFieldValue = new ExtraFieldValue('user');
+        $result = $extraFieldValue->get_item_id_from_field_variable_and_field_value(
+            self::EXTRA_FIELD_OAUTH2_ID,
+            $resourceOwnerId
+        );
+        if (false === $result) {
+            // authenticated user not found in internal database
+            if ('true' !== $this->get(self::SETTING_CREATE_NEW_USERS)) {
+                throw new RuntimeException(get_lang('no_user_has_this_oauth_code'));
```
Comment (jmontoyaa) [inline plugin/oauth2/src/OAuth2.php:184]: Check the use of get_lang in this PR

File: plugin/oauth2/src/OAuth2.php [hunk 17]
```diff
@@ -0,0 +1,382 @@
...
+        $response = $provider->getParsedResponse($request);
+        if (false === is_array($response)) {
+            throw new UnexpectedValueException(get_lang('invalid_json_received_from_provider'));
+        }
+        $resourceOwnerId = $this->getValueByKey(
+            $response,
+            $this->get(self::SETTING_RESPONSE_RESOURCE_OWNER_ID)
+        );
+        if (empty($resourceOwnerId)) {
+            throw new RuntimeException(get_lang('wrong_response_resource_owner_id'));
```
Comment (jmontoyaa) [inline plugin/oauth2/src/OAuth2.php:174]: Variable inside get_lang() should be in PascalCase Example: 

`get_lang('WrongResponseResourceOwnerId');`

But it seems that this variable it's a plugin variable so, it should be 

`$this->get_lang('WrongResponseResourceOwnerId')`

**Comment by jmontoyaa** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 9
           

Complexity increasing per file
==============================
- plugin/oauth2/src/OAuth2.php  11
- main/inc/lib/online.inc.php  3
- plugin/oauth2/config.dist.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956&bid=18593000)

File: plugin/oauth2/lang/english.php [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
```
Comment (jmontoyaa) [inline plugin/oauth2/lang/english.php:1]: Codacy found an issue: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/plugin.php (hunk 9 above)
Comment (jmontoyaa) [inline plugin/oauth2/plugin.php:1]: Codacy found an issue: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/install.php (hunk 1 above)
Comment (jmontoyaa) [inline plugin/oauth2/install.php:1]: Codacy found an issue: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/index.php (hunk 4 above)
Comment (jmontoyaa) [inline plugin/oauth2/index.php:1]: Codacy found an issue: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/src/callback.php (hunk 8 above)
Comment (jmontoyaa) [inline plugin/oauth2/src/callback.php:1]: Codacy found an issue: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/src/OAuth2.php [hunk 19]
```diff
@@ -0,0 +1,381 @@
+<?php
```
Comment (jmontoyaa) [inline plugin/oauth2/src/OAuth2.php:1]: Codacy found an issue: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/login.php (hunk 2 above)
Comment (jmontoyaa) [inline plugin/oauth2/login.php:1]: Codacy found an issue: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: main/auth/external_login/login.oauth2.php (hunk 3 above)
Comment (jmontoyaa) [inline main/auth/external_login/login.oauth2.php:1]: Codacy found an issue: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

File: plugin/oauth2/config.dist.php (hunk 5 above)
Comment (jmontoyaa) [inline plugin/oauth2/config.dist.php:1]: Codacy found an issue: [End of line character is invalid; expected "'\n'" but found "\n"](https://app.codacy.com/manual/chamilo/chamilo-lms/pullRequest?prid=5679956)

**Pull request merged**